independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/OpenContainerPathTree.java

```diff
@@ -0,0 +1,189 @@
+package io.quarkus.paths;
+
+import java.nio.file.FileSystem;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.regex.Pattern;
+
+public abstract class OpenContainerPathTree extends PathTreeWithManifest implements OpenPathTree {
+
+    private static final boolean USE_WINDOWS_ABSOLUTE_PATH_PATTERN = !FileSystems.getDefault().getSeparator().equals("/");
+
+    private static volatile Pattern windowsAbsolutePathPattern;
+
+    private static Pattern windowsAbsolutePathPattern() {
+        return windowsAbsolutePathPattern == null ? windowsAbsolutePathPattern = Pattern.compile("[a-zA-Z]:\\\\.*")
+                : windowsAbsolutePathPattern;
+    }
+
+    static boolean isAbsolutePath(String path) {
+        return path != null && !path.isEmpty()
+                && (path.charAt(0) == '/' // we want to check for '/' on every OS
+                        || USE_WINDOWS_ABSOLUTE_PATH_PATTERN
+                                && (windowsAbsolutePathPattern().matcher(path).matches())
+                        || path.startsWith(FileSystems.getDefault().getSeparator()));
+    }
+
+    static void ensureResourcePath(FileSystem fs, String path) {
+        if (isAbsolutePath(path)) {
+            throw new IllegalArgumentException("Expected a path relative to the root of the path tree but got " + path);
+        }
+        // this is to disallow reading outside the path tree root
+        if (path != null && path.contains("..")) {
+            for (Path pathElement : fs.getPath(path)) {
+                if (pathElement.toString().equals("..")) {
+                    throw new IllegalArgumentException("'..' cannot be used in resource paths, but got " + path);
+                }
+            }
+        }
+    }
+
+    protected PathFilter pathFilter;
+
+    /**
+     * For deserialization of DirectoryPathTree
+     */
+    public OpenContainerPathTree() {
+    }
+
+    protected OpenContainerPathTree(PathFilter pathFilter) {
+        this(pathFilter, false);
+    }
+
+    protected OpenContainerPathTree(PathFilter pathFilter, boolean manifestEnabled) {
+        super(manifestEnabled);
+        this.pathFilter = pathFilter;
+    }
+
+    protected OpenContainerPathTree(PathFilter pathFilter, PathTreeWithManifest pathTreeWithManifest) {
+        super(pathTreeWithManifest);
+        this.pathFilter = pathFilter;
+    }
+
+    /**
+     * This is the path to the container.
+     * <p>
+     * In the case of a zip archive, it's the path of the archive.
+     * In the case of a directory, it's the directory.
```

Comment:
There is a bit of an issue with `MultiRootPathTree` here. That's what `getRoots()` currently returns.
`MultiRootPathTree` should get another thought,